source/extensions/outlier_detection_monitors/common/monitor_base_impl.h

```diff
@@ -0,0 +1,199 @@
+class HTTPCodesBucket : public TypedErrorsBucket<Upstream::Outlier::ExtResultType::HTTP_CODE> {
+public:
+  HTTPCodesBucket() = delete;
+  HTTPCodesBucket(uint64_t start, uint64_t end) : start_(start), end_(end) {}
+  bool matches(const TypedExtResult<Upstream::Outlier::ExtResultType::HTTP_CODE>&) const override;
+
+  virtual ~HTTPCodesBucket() {}
+
+private:
+  uint64_t start_, end_;
+};
+
+// Class defines a "bucket" which catches LocalOriginEvent.
+class LocalOriginEventsBucket
+    : public TypedErrorsBucket<Upstream::Outlier::ExtResultType::LOCAL_ORIGIN> {
+public:
+  LocalOriginEventsBucket() = default;
+  bool
+  matches(const TypedExtResult<Upstream::Outlier::ExtResultType::LOCAL_ORIGIN>&) const override;
+};
+
+// Class groups error buckets. Buckets may be of different types.
+// Base class for various types of monitors.
+// Each monitor may implement different health detection algorithm.
+class Monitor {
+public:
+  Monitor(const std::string& name, uint32_t enforce) : name_(name), enforce_(enforce) {}
+  Monitor() = delete;
+  virtual ~Monitor() {}
+  void reportResult(const ExtResult&);
+
+  void
+  setCallback(std::function<void(uint32_t, std::string, absl::optional<std::string>)> callback) {
+    callback_ = callback;
+  }
+
+  void reset() { onReset(); }
+  std::string name() const { return name_; }
+
+  void processBucketsConfig(
+      const envoy::extensions::outlier_detection_monitors::common::v3::ErrorBuckets& config);
+  void addErrorBucket(ErrorsBucketPtr&& bucket) { buckets_.push_back(std::move(bucket)); }
+
+protected:
+  virtual bool onError() PURE;
+  virtual void onSuccess() PURE;
+  virtual void onReset() PURE;
+  // Default extra info is empty string. Descendant classes may overwrite it.
+  virtual std::string getFailedExtraInfo() { return ""; }
+
+  std::string name_;
+  uint32_t enforce_{100};
+  std::vector<ErrorsBucketPtr> buckets_;
+  std::function<void(uint32_t, std::string, absl::optional<std::string>)> callback_;
+};
+
+using MonitorPtr = std::unique_ptr<Monitor>;
+
+class MonitorsSet {
+public:
+  void addMonitor(MonitorPtr&& monitor) { monitors_.push_back(std::move(monitor)); }
+  const std::vector<MonitorPtr>& monitors() { return monitors_; }
+
+private:
+  std::vector<MonitorPtr> monitors_;
+};
+
+class MonitorFactoryContext {
+public:
+  MonitorFactoryContext(ProtobufMessage::ValidationVisitor& validation_visitor)
+      : validation_visitor_(validation_visitor) {}
+  ProtobufMessage::ValidationVisitor& messageValidationVisitor() { return validation_visitor_; }
+
+private:
+  ProtobufMessage::ValidationVisitor& validation_visitor_;
+};
+
+class MonitorFactory : public Envoy::Config::TypedFactory {
+public:
+  ~MonitorFactory() override = default;
+
+  virtual MonitorPtr createMonitor(const std::string& name, const Protobuf::Message& config,
+                                   MonitorFactoryContext& context) PURE;
+
+  std::string category() const override { return "envoy.outlier_detection_monitors"; }
+};
```

Comment:
Hmm. I am confused. I thought that `outlier_detection.h` is mostly to define abstract classes. Are you sure that `outlier_detection.h` is better location? I was under impression that `source/extensions/...` should hide as many details as possible.

Reply:
Yeah. It would be great to hide more detail in the `/source/extensions/...`. But this factory will be used in the `/source/common/upstream` and typically is part of the core code.

More important point is that we should avoid the dependency from common -> extensions.

Reply:
Done. I had to move few classes to `outlier_detection.h` and this allowed me to not to `include source/extensions/...`.